### DISCUSSION TITLE
eap-radius ID Verification issue

### DISCUSSION THREAD

**Discussion by sangdrax8** [strongswan/strongswan#453] [Q&A] [answered]
I have recently found an issue with my eap-radius authentication.  It seems my server is not validating the identity the client is sending with any fields in the certificate.  I only found this when using a certificate which was already working (MacOS), with a strongswan client (Ubuntu).  The client was rejecting the certificate because it didn't match the ID that was being configured.  I re-created the certificate to correctly contain the ID needed in a subjectAltName, and everything worked right away.  However this raised the question of why the cert still worked on the Mac.

I went back to the MacOS VPN client (which I was using previously) and discovered that it does not enforce this.  I can simply change the Identifier to another user, and since the certificate is valid (signed by the correct CA) and the user does exist, I can still connect.  Seeing as how I can't trust client side applications for this sort of thing, it would make sense if I could somehow get strongswan on the server side to validate this.  I am guessing I must not fully understand how eap-radius works, and which piece is the issue here.  Since my strongswan client on Linux noticed the issue, I am hopeful that the server can be configured to do the same?

Currently I am running:
Strongswan: 5.9.2,  Linux 4.19.0 (debian).

swanctl.conf

```
con-mobile-defaults {
        fragmentation = yes
        unique = replace
        version = 2
        proposals = aes256-sha256-modp8192,aes256gcm128-sha256-modp1024,aes256-sha256-modp2048
        dpd_delay = 10s
        dpd_timeout = 60s
        rekey_time = 25920s
        reauth_time = 0s
        over_time = 2880s
        rand_time = 2880s
        encap = no
        mobike = yes
        local_addrs = 192.168.0.48
        remote_addrs = 0.0.0.0/0,::/0
        pools = mobile-pool-v4, radius
        send_cert = always
        local {
                id = 192.168.0.48
                auth = pubkey
                cert {
                        file = /etc/swanctl/x509/server.crt
                }
        }
        remote {
                id = %any
                eap_id = %identity
                auth = eap-radius
        }
        children {
                con-mobile {
                        dpd_action = clear
                        mode = tunnel
                        policies = yes
                        life_time = 3960s
                        rekey_time = 3600s
                        rand_time = 300s
                        start_action = none
                        local_ts = 0.0.0.0/0
                        esp_proposals = aes128gcm128,aes128gcm96,aes128gcm64,aes256-sha1,aes256-sha256
                }
        }
}
pools {
        mobile-pool-v4 : mobile-pool {
                addrs = 192.168.10.0/24
        }
}
mobile-pool { 
        dns = 192.168.0.15,192.168.0.16
}
secrets {
        private-1 {
                file = server.key
        }
}
```

**Comment by sangdrax8** [comment]
Although, I would still be curious if this is something that I have incorrectly configured on the strongswan side, I have found a fix for right now which solves my initial issue.  Freeradius does have an option to compare the user name to the CN of the certificate provided.  I have activated this and at least prevented the issue which started this.

I am still curious though, when looking into the issues with my strongswan client rejecting my cert/ID combination, it was very clear that the ID had to match the entire subject, or a subjectAltName.  The above radius fix doesn't actually do that... which makes me wonder if there is a piece here that is not actually standards compliant, even if it does work as intended now.  Technically I still have users who CAN log in with an ID which does not match the full subject, OR a subjectAltName, but a piece of the subject (CN).

**Reply by gsra99** [reply]
I get the following error after upgrading the client Linux machine to StrongSwan 5.9.2. I have removed my server's ip address.

> [IKE] server requested EAP_TLS authentication (id 0x01)
[ENC] generating IKE_AUTH request 2 [ EAP/RES/TLS ]
[NET] sending packet: from 192.168.1.124[4500] to XX.XX.XX.XX[4500] (348 bytes)
[NET] received packet: from XX.XX.XX.XX[4500] to 192.168.1.124[4500] (76 bytes)
[ENC] parsed IKE_AUTH response 2 [ EAP/FAIL ]
[IKE] received EAP_FAILURE, EAP authentication failed

I would be grateful to know if this is the error you got on the client.

**Reply by tobiasbrunner** [reply]
You have to read the logs of the VPN and RADIUS servers to see what the actual problem as the only thing the client receives is an error notification, not the actual reason for the error.

**Reply by gsra99** [reply]
@tobiasbrunner On the radius server I get the following message both for the Linux Strongswan 5.9.2 client and Android 2.3.3 client.

> (1) # Executing group from file /etc/raddb/sites-enabled/default
(1)   authenticate {
(1) eap: Expiring EAP session with state 0xd1555205d1545fbb
(1) eap: Finished EAP session with state 0xd1555205d1545fbb
(1) eap: Previous EAP request found for state 0xd1555205d1545fbb, released from the list
(1) eap: Peer sent packet with method EAP TLS (13)
(1) eap: Calling submodule eap_tls to process data
(1) eap_tls: Continuing EAP-TLS
(1) eap_tls: Peer indicated complete TLS record size will be 194 bytes
(1) eap_tls: Got complete TLS record (194 bytes)
(1) eap_tls: [eaptls verify] = length included
(1) eap_tls: (other): before SSL initialization
(1) eap_tls: TLS_accept: before SSL initialization
(1) eap_tls: TLS_accept: before SSL initialization
(1) eap_tls: <<< recv TLS 1.3  [length 00bd] 
(1) eap_tls: >>> send TLS 1.2  [length 0002] 
(1) eap_tls: ERROR: TLS Alert write:fatal:handshake failure
tls: TLS_accept: Error in error
(1) eap_tls: ERROR: Failed in __FUNCTION__ (SSL_read): error:14201076:SSL routines:tls_choose_sigalg:no suitable signature algorithm
(1) eap_tls: ERROR: System call (I/O) error (-1)
(1) eap_tls: ERROR: TLS receive handshake failed during operation
(1) eap_tls: ERROR: [eaptls process] = fail
(1) eap: ERROR: Failed continuing EAP TLS (13) session.  EAP sub-module failed
(1) eap: Sending EAP Failure (code 4) ID 1 length 4
(1) eap: Failed in EAP select
(1)     [eap] = invalid
(1)   } # authenticate = invalid
(1) Failed to authenticate the user
(1) Using Post-Auth-Type Reject

However, using the Linux Strongswan 5.9.1 client and Android 2.3.2 client the EAP-TLS authentication works without any issues, so I have downgraded the version on my Linux laptop client and Android mobile phone.

**Reply by tobiasbrunner** [reply]
> (1) eap_tls: <<< recv TLS 1.3 [length 00bd]

Hm, this looks weird. TLS 1.3 is not enabled by default, so not sure why this would get logged (especially since the TLS version in the header is NEVER 1.3). I'd be interested in seeing a PCAP file of this interaction with FreeRADIUS (captured on the RADIUS server).

> (1) eap_tls: ERROR: Failed in **FUNCTION** (SSL_read): error:14201076:SSL routines:tls_choose_sigalg:no suitable signature algorithm

With the addition of TLS 1.3 with 5.9.2, there were a lot of other changes to our TLS stack. One was removing legacy algorithms and cipher suites (in particular, signature schemes with MD5 and SHA-1 are not proposed anymore). What versions of FreeRADIUS and OpenSSL are you using? Are those maybe restricted to legacy ciphers?

**Reply by gsra99** [reply]
Thanks @tobiasbrunner, I was using SHA1 signatures for my certificates, so reissued new certificates using SHA512 signatures and have no issues with connecting using Strongswan 5.9.2 Linux clients or Strongswan 2.3.3 Android clients.

**Comment by tobiasbrunner** [comment, marked as answer]
If you use EAP-TLS via _eap-radius_ plugin, strongSwan is not involved in the actual authentication (i.e. the validation of the certificate chain or identities etc.). So as you saw, you have to get your RADIUS server to enforce this (or use the _eap-tls_ plugin to terminate the authentication on the VPN gateway but you'd loose e.g. the _radius_ pool functionality that way, you still could do RADIUS Accounting, though).

Technically, the EAP-Identity is an arbitrary value that a RADIUS server can use to e.g. to select configs. The [EAP-TLS RFC](https://datatracker.ietf.org/doc/html/rfc5216#section-2.2) explicitly says this:

> Since the identity presented in the EAP-Response/Identity need not be related to the identity presented in the peer certificate, EAP-TLS implementations SHOULD NOT require that they be identical.  However, if they are not identical, the identity presented in the EAP-Response/Identity is unauthenticated information, and SHOULD NOT be used for access control or accounting purposes.

So strongSwan's _eap-tls_ plugin is stricter here than recommended by always enforcing that the EAP-Identity is confirmed by the certificate (i.e. either matches the full subject DN or one of the SANs, that's how it's done in IKE, so we do the same for TLS, we don't do partial DN matches like many TLS implementations).

On the other hand, if a RADIUS server does not enforce this, the EAP-Identity value should not be used in strongSwan. Unfortunately, as far as I know, we don't get that information back from the RADIUS server. So it's probably best to ensure that the RADIUS server enforces this if you use the identity for anything in strongSwan.

**Reply by sangdrax8** [reply]
Thank you!  Now I feel like I not only have a solid config, but that I didn't miss configure in strongswan setup either.
